[SYCL][ESIMD] Provide the for_all_combinations utility

yuriykoch opened the pull request:
Provides a generic way to represent a combinations of types and compile-time
values to cover. Supports any number of type and value packs to iterate over,
with any combination of named and unnamed type packs and value packs
possible

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Move getters to the end of type_coverage.hpp

To ensure more complex logic resides nearby each other

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Provide the for_all_combinations utility

Provides a generic way to represent a combinations of types and compile-time
values to cover. Supports any number of type and value packs to iterate over

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

vasilytric commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -161,10 +164,10 @@ int main(int, char **) {
   const auto dims = get_all_dimensions();
 
-  passed &= for_all_types_and_dims<run_test, initializer>(types, dims, queue);
-  passed &= for_all_types_and_dims<run_test, var_decl>(types, dims, queue);
-  passed &= for_all_types_and_dims<run_test, rval_in_expr>(types, dims, queue);
-  passed &= for_all_types_and_dims<run_test, const_ref>(types, dims, queue);
+  passed &= for_all_combinations<run_test, initializer>(types, dims, queue);
+  passed &= for_all_combinations<run_test, var_decl>(types, dims, queue);
+  passed &= for_all_combinations<run_test, rval_in_expr>(types, dims, queue);
+  passed &= for_all_combinations<run_test, const_ref>(types, dims, queue);
```
What we will win by this patch set?
According to my understanding we should will decrease the numbed of function calls, but function names were changed (from `for_all_types_and_dims` to `for_all_combinations`) and other architectures changes

yuriykoch replied:
By any means, the `4`-lines refactoring doesn't worth the `+200 -100` lines patch, does it? Currently there is nothing worth to show with such simple tests. Still it will fit well with conversion constructor and with much more complex tests, right?

As we discussed, I'm a bit worried about renaming the `get_description()` methods to the `to_string()` methods, as it's closely related to the logs architecture. Let me have a second thought about it.

yuriykoch replied:
As Vasily suggested (thanks Vasily), it might be useful to provide example using the tests on the `fill` constructor merged recently. By doing so I've found some unwanted architecture restrictions, so now there is support for unnamed type pack.

Addressed in commit 5acd8257deb29552b327cdef567f7f509eb2309f

yuriykoch committed:
[SYCL][ESIMD] Fix recursion for linux

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Fix clang-format

The newer version should be used instead of clang-format-10

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
Merge branch 'intel' into feature/for_all_combinations

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Add unnamed type pack support

By doing so we avoid additional restriction on architecture of
the test case functors and on logging architecture, as we don't
need to construct string description for every type/value on the
top level.
We are still free to use named type packs, giving us a possibility
to have any combination of named packs and unnamed packs, with value
packs supported in both variations.
Tests are updated as there is no more for_all_types_and_dims function

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
[SYCL][ESIMD] Fix clang-format

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch committed:
Merge branch 'intel' into feature/for_all_combinations

To enable CI verification for the ctor_copy test

Signed-off-by: Kochetkov, Yuriy <yuriyx.kochetkov@intel.com>

yuriykoch commented:
@sndmitriev Would you mind to take a look?

yuriykoch committed:
Merge branch 'intel' into feature/for_all_combinations

yuriykoch commented:
Merge conflicts resolved. @sndmitriev Friendly ping

yuriykoch commented:
@kbobrovs PR is ready to merge
